app/jobs/regular/grant_badge.rb

```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+module Jobs
+  class GrantBadge < ::Jobs::Base
+    def execute(args)
+      return unless SiteSetting.enable_badges
+
+      badge = Badge.enabled.find_by(id: args[:badge_id])
+      return unless badge
+
+      begin
+        BadgeGranter.backfill(badge)
+      rescue => ex
+        # TODO - expose errors in UI
+        Discourse.handle_job_exception(
+          ex,
+          error_context({}, code_desc: "Exception granting badges", extra: { badge_id: badge.id }),
+        )
+      end
+
+      # If this instance is the last job to be processed, schedule the
+      # EnsureBadgeConsistency job. This guarantees it runs only once after all
+      # badges have been granted.
+      if Discourse.redis.decr("grant_badge_remaining") <= 0
+        Discourse.redis.del("grant_badge_remaining")
+        Jobs.enqueue(:ensure_badge_consistency)
+      end
```

Comment:
I think this will work but I'm still not a fan of this approach because we are arguably introducing some sort of coupling between two jobs.